docs/src/rfc/text/0017-aml-parser.md

```diff
@@ -0,0 +1,331 @@
+and can be reused partially *from* the existing Rust `acpi` crate.
+
+This proposed `patina-acpi` utility crate crate shares functionality with the Rust  `os-dev acpi` crate; as such,
+it remains to be seen if collaboration is possible to expose this shared parsing functionality,
+or if `patina-acpi` will borrow/fork from `acpi`.
+
+### AML Handles
+
+By spec definition, an AML handle is an opaque handle returned from opening a DSDT or SSDT,
+on which AML traversal and patching operations can be performed.
+
+Internally, each `AmlHandle` object, aliased as `AmlSdtHandleInternal`,
+represents a cursor within an AML stream. Each handle can be conceptualized as a "node"
+in the AML object tree of parent-child-relationships.
+
+```rust
+pub(crate) struct AmlSdtHandleInternal {
+    table_key: TableKey,
+    offset: usize,
+    size: usize,
+    modified: bool,
+    byte_encoding: AmlByteEncoding,
+    parent_end: Option<usize>,
+}
+
+impl AmlSdtHandleInternal {
+    fn new(table_key: TableKey, offset: usize, size: usize) -> Self {
+        Self {
+            table_key,
+            offset,
+            size,
+            modified: false,
+            byte_encoding: AmlByteEncoding::default(),
+            parent_end: None,
+        }
+    }
+}
+
+pub type AmlHandle = AmlSdtHandleInternal;
+```
+
+The `size` of an `AmlSdtHandleInternal` refers to its full size, including any children (`TermList`).
+
+The `offset` refers to its offset with the AML stream of the table.
+Offset 0 is the start of the AML stream, and the highest offset is at the end of `table_length`.
+
+`modified` ensures the corresponding table (which can be retrieved through `table_key`) has an updated checksum
+if the contents are modified by `set_option`.
+
+Each handle stores the `parent_end` of its parent node, which is the parent's `size` + `offset` (useful for retrieving siblings).
+
+Each handle also stores its own byte encoding, specifying its own layout:
+
+```rust
+#[derive(PartialEq, Eq, Hash)]
+struct AmlByteEncoding {
+    opcode: AmlOpcode,
+    operands: Vec<AmlOperand>,
+    attributes: AmlOpAttributes,
+}
+
+bitflags! {
+    pub struct AmlOpAttributes: u32 {
+        /// If opcode has a pkg_length field.
+        const HAS_PKG_LENGTH  = 0x0000_0001;
+
+        /// If opcode has children.
+        const HAS_CHILD_OBJ   = 0x0000_0002;
+    }
+}
+
+/// Represents the possible opcodes. 
+pub enum AmlOpcode {
+    BaseOp(BaseOpcode),
+    ExtOp(ExtOpcode),
+}
+
+pub enum BaseOpcode {
+    ZeroOp,
+    AliasOp,
+    ...
+}
+
+pub enum ExtOpcode {
+    MutexOp,
+    EventOp,
+    ...
+}
+
+pub enum AmlOperand {
+    Opcode(AmlOpcode),
+    Name(AmlNameString), // Represents a NameString (AML path). Not to be confused with a string literal 
+    ...
+}
+```
+
+### AML Trait Interface
+
+The `AmlParser` service generally derives from the ACPI SDT protocol, and allows for traversal of the AML object tree.

```

Comment:
It's been mentioned elsewhere, but I'm not sure that the protocol interface should serve as the primary design point. That was a PI-spec C interface. For the purpose of the RFC, I'd consider inverting this. Define an independent, idiomatic Rust service interface and have the protocol be a wrapper around that.